### PR TITLE
pr-pull: allow to pull from multiple workflows

### PR DESCRIPTION
**Pull request by iMichka**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

### CODE REVIEW

**Comment by iMichka** [comment]
I tested this locally and it does download the Mac and Linux bottles for https://github.com/Homebrew/homebrew-core/pull/65417

I am a little bit unsure where the code is that creates the bottle commit though, to check if it will add the linux bottle too.

**Comment by iMichka** [comment]
I think I should also check if the wheezy workflow has run before pulling from it.

**Comment by iMichka** [comment]
I re-worked this to be much more generic.

I removed the --workflow flag. This means that as soon as this is merged, we need to adapt all our workflows. Should I keep it for backward compatibility?

**Comment by reitermarkus** [comment]
> I removed the --workflow flag.

I think we should keep both.

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 1]
```diff
@@ -62,6 +63,11 @@ def pr_pull_args
       flag   "--bintray-mirror=",
              description: "Use the specified Bintray repository to automatically mirror stable URLs "\
                           "defined in the formulae (default: `mirror`)."
+      comma_array "--workflows=",
```
Comment (reitermarkus) [inline Library/Homebrew/dev-cmd/pr-pull.rb:66]: Does 

```suggestion
      comma_array "--workflow=", "--workflows=",
```

work?

Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb:66]: Wrong number of arguments. comma_array expects only 1 argument.

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 2]
```diff
@@ -50,7 +50,8 @@ def pr_pull_args
              depends_on:  "--autosquash",
              description: "Message to include when autosquashing revision bumps, deletions, and rebuilds."
       flag   "--workflow=",
-             description: "Retrieve artifacts from the specified workflow (default: `tests.yml`)."
+             description: "Retrieve artifacts from the specified workflow (default: `tests.yml`). "\
+                          "Legacy: use --workflows instead"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:54]: Could this be `odeprecated`?

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 3]
```diff
@@ -357,7 +363,11 @@ def download_artifact(url, dir, pr)
   def pr_pull
     args = pr_pull_args.parse
 
-    workflow = args.workflow || "tests.yml"
+    workflows = if args.workflow.nil?
+      args.workflows || ["tests.yml"]
+    else
+      [args.workflow] || ["tests.yml"]
+    end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:370]:
```suggestion
    workflows = if args.workflow.blank?
      args.workflows.presence || ["tests.yml"]
    else
      [args.workflow].compact.presence || ["tests.yml"]
    end
```

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 4]
```diff
@@ -401,8 +411,22 @@ def pr_pull
             next
           end
 
-          url = GitHub.get_artifact_url(user, repo, pr, workflow_id: workflow, artifact_name: artifact)
-          download_artifact(url, dir, pr)
+          workflows.each do |workflow|
+            workflow_run = GitHub.get_workflow_run(
+              user, repo, pr, workflow_id: workflow, artifact_name: artifact
+            )
+            if !args.ignore_missing_artifacts.nil? &&
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:418]:
```suggestion
            if args.ignore_missing_artifacts.present? &&
```

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 5]
```diff
@@ -62,6 +63,11 @@ def pr_pull_args
       flag   "--bintray-mirror=",
              description: "Use the specified Bintray repository to automatically mirror stable URLs "\
                           "defined in the formulae (default: `mirror`)."
+      comma_array "--workflows=",
+                  description: "Retrieve artifacts from the specified workflow (default: `tests.yml`) "\
+                               "Comma-separated list to include multiple workflows."
+      comma_array "--ignore-missing-artifacts=",
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:69]: Can you explain what this is used for in this case? I don't really fully understand.

Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb:69]: Yes, sorry it's a little bit hard to follow and we did not come up with a better flag name.

The wheezy_tests.yml workflow is only run for 2 files, patchelf and binutils. Else it is not run at all.

This means that we want to write the following in homebrew-core:
`brew pr-pull --workflows=tests.yml,wheezy_tests.yml --ignore-missing-artifacts=wheezy_tests.yml`

This means: fetch the bottles for both workflows, but if the second one is not run no need to fetch the bottles.

There is also a final check in `get_artifact_url` if the workflow has run (in case you pull a running or failed workflow).

The whole setup is a little bit weird but it is because we can not have conditionals in the GitHub workflow .yml files.
I tried to keep the code as generic as possible.

There will soon be a third workflow for linux bottles, and maybe a fourth for self-hosted runners.

Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/pr-pull.rb:69]: > This means that we want to write the following in homebrew-core:
> `brew pr-pull --workflows=tests.yml,wheezy_tests.yml --ignore-missing-artifacts=wheezy_tests.yml`

Is this run by a human or @BrewTestBot? If a human: it feels like having a dedicated flag for all this is going to be less code complexity and less error-prone than remembering the exact invocation.

> There will soon be a third workflow for linux bottles, and maybe a fourth for self-hosted runners.

We'll need to figure out a way of merging this workflows, I think.

Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb:69]: By the bot, here: https://github.com/Homebrew/homebrew-core/blob/master/.github/workflows/publish-commit-bottles.yml#L41

> We'll need to figure out a way of merging this workflows, I think.

I already discussed this with @dawidd6 and there seem to be no easy solution yet. We might have some code duplication in workflows for some time.

Comment (iMichka) [reply Library/Homebrew/dev-cmd/pr-pull.rb:69]: I am seeing all this as a sort of V1, to get things going. There are still things to iron out, but if I remember well it took a few months to get the current workflows setup as we have them now. Things will evolve (I hope).

**Review by MikeMcQuaid — APPROVED** [review]
Would like to see further refactoring on this but it seems fine for now.
